panel/core/esm.py

```diff
@@ -0,0 +1,148 @@
+from __future__ import annotations
+
+import os
+import pathlib
+import textwrap
+
+from typing import (
+    TYPE_CHECKING, Any, ClassVar, Dict, List, Literal, Mapping, Optional,
+    Tuple, Type,
+)
+
+import jinja2
+import param
+
+from param.parameterized import ParameterizedMetaclass
+
+from .io.datamodel import construct_data_model
+from .io.state import state
+from .models import ReactiveESM as _BkReactiveESM
+from .reactive import Reactive, ReactiveCustomBase, ReactiveMetaBase
+from .viewable import Layoutable, Viewable
+
+if TYPE_CHECKING:
+    from bokeh.document import Document
+    from bokeh.model import Model
+    from pyviz_comms import Comm
+
+
+class ReactiveESMMetaclass(ReactiveMetaBase):
+
+    def __init__(mcs, name: str, bases: Tuple[Type, ...], dict_: Mapping[str, Any]):
+        mcs.__original_doc__ = mcs.__doc__
+        ParameterizedMetaclass.__init__(mcs, name, bases, dict_)
+
+        # Create model with unique name
+        ReactiveMetaBase._name_counter[name] += 1
+        model_name = f'{name}{ReactiveMetaBase._name_counter[name]}'
+        mcs._data_model = construct_data_model(
+            mcs, name=model_name, ignore=list(Reactive.param)
+        )
+
+
+class ReactiveESM(ReactiveCustomBase, metaclass=ReactiveESMMetaclass):
+
+    _esm: ClassVar[str | os.PathLike] = ""
+    _import_map: ClassVar[Dict[str, Dict[Literal['imports', 'scopes'], str]]] = {}
+
+    __abstract = True
+
+    def __init__(self, **params):
+        super().__init__(**params)
+        self._watching_esm = False
+
+    def _render_esm(self):
+        if isinstance(self._esm, pathlib.PurePath):
+            esm = self._esm.read_text()
+        else:
+            esm = self._esm
+        esm = textwrap.dedent(esm)
+        template = jinja2.Template(esm)
+        return template.render({})
+
+    async def _watch_esm(self):
+        import watchfiles
```

Comment:
I believe you cannot assume watchfiles is installed. Instead you should check if it is or add error handling.